sw/device/lib/crypto/drivers/entropy_kat.c

```diff
@@ -0,0 +1,169 @@
+ */
+typedef enum entropy_csrng_internal_state_id {
+  /**
+   * CSRNG instance assigned to Entropy Distribution Network (EDN) 0.
+   */
+  kCsrngInternalStateIdEdn0 = 0,
+  /**
+   * CSRNG instance assigned to Entropy Distribution Network (EDN) 1.
+   */
+  kCsrngInternalStateIdEdn1 = 1,
+  /**
+   * CSRNG instance assigned to software interface.
+   */
+  kCsrngInternalStateIdSw = 2,
+} entropy_csrng_internal_state_id_t;
+
+/**
+ * CSRNG internal state.
+ */
+typedef struct entropy_csrng_internal_state {
+  /**
+   * Indicates the number of requests for pseudorandom bits since instantiation
+   * or reseeding.
+   */
+  uint32_t reseed_counter;
+  /**
+   * Internal V working state with a 128bit block size.
+   */
+  uint32_t v[4];
+  /**
+   * Internal key used to configure the internal CSRNG cipher.
+   */
+  uint32_t key[8];
+  /**
+   * Set to true when the CSRNG instance has been instantiated.
+   */
+  bool instantiated;
+  /**
+   * Set to true when FIPS compliant entropy was provided directly by the
+   * entropy source to instantiate or reseed the CSRNG instance.
+   */
+  bool fips_compliance;
+} entropy_csrng_internal_state_t;
+
+static void entropy_csrng_internal_state_get(
+    entropy_csrng_internal_state_id_t instance_id,
+    entropy_csrng_internal_state_t *state) {
+  // Select the instance id to read the internal state from, request a state
+  // machine halt, and wait for the internal registers to be ready to be read.
+  uint32_t reg = bitfield_field32_write(
+      0, CSRNG_INT_STATE_NUM_INT_STATE_NUM_FIELD, instance_id);
+  abs_mmio_write32(kBaseCsrng + CSRNG_INT_STATE_NUM_REG_OFFSET, reg);
+
+  // Read the internal state.
+  state->reseed_counter =
+      abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+
+  for (size_t i = 0; i < ARRAYSIZE(state->v); ++i) {
+    state->v[i] = abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+  }
+
+  for (size_t i = 0; i < ARRAYSIZE(state->key); ++i) {
+    state->key[i] =
+        abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+  }
+
+  uint32_t flags = abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+
+  // The following bit indexes are defined in
+  // https://docs.opentitan.org/hw/ip/csrng/doc/#working-state-values
+  state->instantiated = bitfield_bit32_read(flags, /*bit_index=*/0u);
+  state->fips_compliance = bitfield_bit32_read(flags, /*bit_index=*/1u);
+}
+
+/**
+ * Checks the CSRNG internal state against `expected` values.
+ *
+ * @param csrng A CSRNG handle.
+ * @param expected Expected CSRNG internal state.
+ */
+static status_t check_internal_state(
+    const entropy_csrng_internal_state_t *expected) {
+  entropy_csrng_internal_state_t got = {0};
+  entropy_csrng_internal_state_get(kCsrngInternalStateIdSw, &got);
+  if (memcmp(&got, expected, sizeof(entropy_csrng_internal_state_t)) == 0) {
+    return OK_STATUS();
+  }
+  return INTERNAL();
+}
+
+status_t entropy_csrng_kat(void) {
+  TRY(entropy_csrng_uninstantiate());
```

Comment:
Keep in mind that `TRY` and that `status_ok` (and friends) are not hardened.  If they need to be, we might want to write an alternate implementation of those macros/functions that are hardened.

Also, `status_t` uses one bit to differentiate between good/bad status codes (the sign bit).  The "bit hardness" of the status codes needs to be considered as well.  In the mask ROM, the legacy `kErrorOk` is `0x739`; my PR to change the ROM over to `status_t` preserves this Ok value and all of the flow-control helpers (like `RETURN_IF_ERROR`) still depend on it.  In pseudo-code, they are:

```
let result = some_operation();
if (!(status_ok(result) && result.value == 0x739)) {
    // error case
    return result;
}
// good case, execution continues here
```

Reply:
Captured in #14549 to continue the discussion there.